File: Base/sremu 3/server/trunk/src/Core/Players.cpp
/* 
 * Copyright (C) 2005-2008 SREmu <http://www.sremu.org/>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "GameSocket.h"
#include "ObjectMgr.h"

void GameSocket::SpawnMe() 
{
	CreateSpawnPacket(this);
	Broadcast(false);
}

void GameSocket::DespawnMe() 
{
	Writer.Create(GAME_SERVER_DESPAWN_PLAYER);
		Writer.WriteDWord(Player.General.UniqueID);
	Writer.Finalize();
	Broadcast(false);
}

void GameSocket::SpawnPlayers() 
{
	for(Objects.pIter = Objects.Players.begin(); Objects.pIter != Objects.Players.end(); ++Objects.pIter) 
	{
		if(Objects.pIter->first != Player.General.UniqueID) 
		{
			CreateSpawnPacket(Objects.pIter->second);
			Send(Writer.Buffer, Writer.Size());
		}
	}
}

void GameSocket::CreateSpawnPacket(GameSocket* player) 
{
	Writer.Create(GAME_SERVER_SPAWN);

		Writer.WriteDWord(player->Player.Stats.Model);
		Writer.WriteByte (player->Player.Stats.Volume);
		Writer.WriteByte (player->Player.Stats.Level < 20 ? 1 : 0);
		Writer.WriteByte (0x2D);	// Max item slot

		// Item data here.
		mysqlpp::StoreQueryResult items = 
			db.Query("select * from items where owner=%d and type=0", player->Player.General.CharacterID).store();

		Writer.WriteByte (items.num_rows());
		for(unsigned int i = 0; i < items.num_rows(); i++) 
		{
			Writer.WriteDWord(atoi(items[i]["itemtype"]));
			Writer.WriteByte (atoi(items[i]["plusvalue"]));
		}

		// Avatar data here.
		Writer.WriteByte (4);
		Writer.WriteByte (0);

		Writer.WriteByte (0);
		Writer.WriteDWord(player->Player.General.UniqueID);

		Writer.WriteByte (player->Player.Position.XSector);
		Writer.WriteByte (player->Player.Position.YSector);
		Writer.WriteFloat(player->Player.Position.X);
		Writer.WriteFloat(player->Player.Position.Z);
		Writer.WriteFloat(player->Player.Position.X);
	
		// Angle & movement flags.
		Writer.WriteWord (0);	// Angle
		Writer.WriteByte (0);
		Writer.WriteByte (1);
		Writer.WriteByte (0);
		Writer.WriteWord (0);	// Angle

		Writer.WriteWord (1);
		
		Writer.WriteByte (player->Player.Flags.Berserk);
		Writer.WriteFloat(player->Player.Speeds.WalkSpeed);
		Writer.WriteFloat(player->Player.Speeds.RunSpeed);
		Writer.WriteFloat(player->Player.Speeds.BerserkSpeed);

		Writer.WriteByte (0);	// Number of active buffs

		Writer.WriteWord(strlen((char*)player->Player.General.CharacterName));
		Writer.WriteString(player->Player.General.CharacterName, strlen((char*)player->Player.General.CharacterName));

		// Unknown data.
		Writer.WriteByte (0);
		Writer.WriteByte (1);
		Writer.WriteDWord(0);
		Writer.WriteDWord(0);
		Writer.WriteDWord(0);
		Writer.WriteDWord(0);
		Writer.WriteDWord(0);
		Writer.WriteDWord(0);
		Writer.WriteDWord(0);
		Writer.WriteByte (1);
		Writer.WriteByte (player->Player.Flags.PvP);	// 0 = blue costume pvp flag, 1 = blue costume no flag

	Writer.Finalize();
}

void GameSocket::Broadcast(bool relay, int radius) 
{
	/* This will be implemented much better later. */
	for(Objects.pIter = Objects.Players.begin(); Objects.pIter != Objects.Players.end(); ++Objects.pIter) 
	{
		if(Objects.pIter->first == Player.General.UniqueID && !relay) continue;
		if(radius == 0) 
		{
			Objects.pIter->second->Send(Writer.Buffer, Writer.Size());
		}
	}
}

void GameSocket::OnTarget() 
{
	/* Check object type here. Only players supported now. */
	unsigned int objectid = Reader.ReadDWord();

	Writer.Create(GAME_SERVER_TARGET);

		Objects.pIter = Objects.Players.find(objectid);

		if(Objects.pIter != Objects.Players.end()) 
		{
			// Player.Stats.target = objectid;
			Writer.WriteByte (1);
			Writer.WriteDWord(objectid);
			Writer.WriteDWord(0x10);	// Player object type
			Writer.WriteByte (4);		// Job icon, set appropriately.
		} 
		else 
		{
			/* This shouldn't be happening until we implement monsters/npc */
			//Writer.WriteByte (2);
			Writer.WriteByte (1);
            Writer.WriteDWord(objectid);
			Writer.WriteByte (1);
			Writer.WriteDWord(4);//vida
			Writer.WriteDWord(1);
		}

	Writer.Finalize();
	Send(Writer.Buffer, Writer.Size());
}

void GameSocket::OnIngameNotify() 
{
	/* 1 is sent when going ingame. 
	   5 is sent shortly after.
	   7 is sent after moving the first time. */
	unsigned int id = Reader.ReadDWord();
	switch(id) 
	{
		case 1:
			Objects.Players[Player.General.UniqueID] = this;
			Player.Flags.Ingame = true;
			SpawnMe();
			SpawnPlayers();
			SpawnAllM();//Spawn Auto
			break;
		case 5:
			break;
		case 7:
			break;
	}
}
void GameSocket::OnEmotion()
{
    unsigned char type = Reader.ReadByte();
	Writer.Create(GAME_CLIENT_EMOTION);
    Writer.WriteDWord(Player.General.UniqueID);
	Writer.WriteByte (type);
	Writer.Finalize();
	Broadcast(true);
}
void GameSocket::OnSit()
{
	unsigned char type = Reader.ReadByte();
	
	if(Player.Stats.startx != 4 )
	{
        Player.Stats.startx = type;
    }
	else
	{
	   type = 0;
	   Player.Stats.startx = type;
	}
    Writer.Create(0x3122);
	Writer.WriteDWord(Player.General.UniqueID);
	Writer.WriteByte(1);
	Writer.WriteByte(type);
    Writer.Finalize();
    Broadcast(true);
}
void GameSocket::OnEstrella()
{
	unsigned char type = Reader.ReadByte();
    Writer.Create(0xB683);
	Writer.WriteDWord(Player.General.UniqueID);
	Writer.WriteByte(type);
    Writer.Finalize();
    Broadcast(true);   
}

void GameSocket::OnUpdateSkill()
{ 
	printf("Error Falta: 72CB \n");
}
void GameSocket::OnMastertSkill()
{
    printf("Error Falta: 7165 \n");
}

void GameSocket::OnATSkill()
{
	printf("Error Falta: Skill at \n");
    unsigned char type = Reader.ReadByte();
	int num7;
	float num4;
	float num5;
	switch(type)
	{
	case 1:
		num4 = Player.Position.X;
		num5 = Player.Position.Y;
		Writer.Create(0xB738);
		Writer.WriteDWord(Player.General.UniqueID);
		Writer.WriteByte (1);
		Writer.WriteFloat(Player.Position.XSector);
		Writer.WriteFloat(Player.Position.YSector);
		Writer.WriteFloat(Player.Position.X + num4);
		Writer.WriteFloat(Player.Position.Z);
        Writer.WriteFloat(Player.Position.Z + num5);
		Writer.WriteByte (0);
		Writer.Finalize();
		//Send(Writer.Buffer, Writer.Size());
		break;
	case 2:
		break;


	case 4:
        num7 = Reader.ReadDWord();
		switch(type)
		{
		case 0:
			break;
		case 1:
			break;
		case 4:
			break;

		}
		break;
	}

}

void GameSocket::OnADDStr()
{
	char strx = db.Query("SELECT strength FROM characters WHERE id=%d", Player.General.CharacterID).store().num_rows();
	if(strx != 0)
	{
	   db.Query("UPDATE characters SET strength=strength+1 WHERE id=%d", Player.General.CharacterID).execute();
	   db.Query("UPDATE characters SET attribute=attribute-1 WHERE id=%d", Player.General.CharacterID).execute();
       db.Query("UPDATE characters SET hp=strength*10 WHERE id=%d", Player.General.CharacterID).execute();
       db.Query("UPDATE characters SET cur_hp=strength*10 WHERE id=%d", Player.General.CharacterID).execute();
	   Writer.Create(0xB27A);
	   Writer.WriteByte(1);
	   Writer.Finalize();
       Send(Writer.Buffer, Writer.Size());
	}
	else 
	{
     printf("se cree vivo: %s\n", Player.General.CharacterName);
	}
}
void GameSocket::OnADDInt()
{
	char intx = db.Query("SELECT intelligence FROM characters WHERE id=%d", Player.General.CharacterID).store().num_rows();
    if(intx != 0)
     {
       db.Query("UPDATE characters SET intelligence=intelligence+1 WHERE id=%d", Player.General.CharacterID).execute();
	   db.Query("UPDATE characters SET attribute=attribute-1 WHERE id=%d", Player.General.CharacterID).execute();
       db.Query("UPDATE characters SET mp=intelligence*10 WHERE id=%d", Player.General.CharacterID).execute();
	   db.Query("UPDATE characters SET cur_mp=intelligence*10 WHERE id=%d", Player.General.CharacterID).execute();
	   Writer.Create(0xB552);
	   Writer.WriteByte(1);
	   Writer.Finalize();
       Send(Writer.Buffer, Writer.Size());
	}
	else 
	{
     printf("se cree vivo: %s\n", Player.General.CharacterName);
	}
}

File: Base/Old/Source/GameServer/Game/Objects/Players.cs
/*    <DarkEmu GameServer>
    Copyright (C) <2011>  <DarkEmu>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

using System;
using System.Collections.Generic;
using System.Text;

namespace DarkEmu_GameServer
{
    class Players
    {
        public static int GetObjectIndexAndType(int Index_, uint ObjectId)
        {
            int ObjectIndex = 0;
            Player.Objects[Index_].SelectedObjectType = 0;
            for (int i = 0; i <= Player.PlayersOnline; i++)
            {
                if (Player.General[i].UniqueID == ObjectId && Index_ != i)
                {
                    Player.Objects[Index_].SelectedObjectType = 1;
                    ObjectIndex = i;
                    break;
                }
            }
            if (Player.Objects[Index_].SelectedObjectType != 1)
            {
                for (int j = 0; j <= Monsters.MonsterAmount; j++)
                {
                    if (Monsters.General[j].UniqueID == ObjectId)
                    {
                        Player.Objects[Index_].SelectedObjectType = 2;
                        ObjectIndex = j;
                        break;
                    }
                }
            }
            return ObjectIndex;
        }

        public static int GetObjectIndex( uint ObjectId)
        {
            for (int i = 0; i <= Player.PlayersOnline; i++)
            {
                if (Player.General[i].UniqueID == ObjectId)
                {
                    return i;
                }
            }

            for (int j = 0; j <= Monsters.MonsterAmount; j++)
            {
                if (Monsters.General[j].UniqueID == ObjectId)
                {
                    return j;
                }
            }

            return 0;
        }

        public static void OnTarget(PacketReader reader_, int Index_)
        {
            PacketWriter writer = new PacketWriter();
            uint ObjectId = reader_.ReadDword();

            int ObjectIndex = GetObjectIndexAndType(Index_,ObjectId);                 

            writer.SetOpcode(SERVER_OPCODES.GAME_SERVER_TARGET);
            if (ObjectId == Player.General[Index_].UniqueID)
            {
                writer.AppendWord(0x602);
                ServerSocket.Send(writer.getWorkspace(), Index_);
            }
            else
            {
                writer.AppendByte(1);
                writer.AppendDword(ObjectId);
                switch (Player.Objects[Index_].SelectedObjectType)
                {
                    case 1:
                        writer.AppendDword(0x10);
                        writer.AppendByte(4);
                        break;
                    case 2:
                        writer.AppendByte(1);
                        writer.AppendDword((uint)Monsters.General[ObjectIndex].HP);
                        if (Player.Flags[ObjectIndex].Dead)
                            writer.AppendByte(4);
                        else
                            writer.AppendDword(1);
                        break;
                }

                ServerSocket.Send(writer.getWorkspace(), Index_);
            }

            Player.Objects[Index_].SelectedObjectId = ObjectId;
        }

        public static void OnIngameNotify(PacketReader reader_, int Index_)
        {
            uint id = reader_.ReadDword();
            switch (id)
            {
                case 1:
                    Player.PlayersOnline++;
                    Player.Flags[Index_].Ingame = 1;
                    Character.SpawnMe(Index_);
                    Character.SpawnOtherPlayer(Index_);
                    OnRangeSpawn(Index_);
                    break;
            }
        }

        public static byte[] CreateDeSpawnPacket(uint UniqueId)
        {
            PacketWriter writer = new PacketWriter();
            writer.SetOpcode(SERVER_OPCODES.GAME_SERVER_DESPAWN);
            writer.AppendDword(UniqueId);
            return writer.getWorkspace();
        }


        public static void OnRangeSpawn(int Index_)
        {
            for (int i = 0; i < Monsters.MonsterAmount; i++)
            {
                if (Monsters.General[i].UniqueID != 0 && !Monsters.General[i].Dead)
                {
                    double Distance = Formula.CalculateDistance(Player.Position[Index_], Monsters.Position[i]);
                    if (Distance <= 800)
                    {
                        if (!Player.Objects[Index_].SpawnedMonsterIndex.Contains(i))
                        {
                            ServerSocket.Send(MonsterSpawn.CreateSpawnPacket(Monsters.General[i].ID, Monsters.General[i].UniqueID, Monsters.Position[i], Monsters.General[i].Type), Index_);
                            Player.Objects[Index_].SpawnedMonsterIndex.Add(i);
                        }
                    }
                }
            }
            for (int i = 0; i < Item.ItemAmount; i++)
            {
                if (Item.General[i].UniqueID != 0)
                {
                    double Distance = Formula.CalculateDistance(Player.Position[Index_], Item.Position[i]);
                    if (Distance <= 800)
                    {
                        if (!Player.Objects[Index_].SpawnedItemsIndex.Contains(i))
                        {
                            ServerSocket.Send(Items.CreateSpawnPacket(Item.General[i], Item.Position[i]), Index_);
                            Player.Objects[Index_].SpawnedItemsIndex.Add(i);
                        }
                    }
                }
            }
        }
    }
}

